### ISSUE TITLE
Class Diagram - Prepare a diagram for Achievements

### ISSUE THREAD

**Issue by hsuleerkul** [bounswe/bounswe2022group3#72]
As mentioned in [meeting notes](https://github.com/bounswe/bounswe2022group3/wiki/Meeting-Notes-%236-01.04.2022), `Achievements` class should be described. In this diagram, the first version of achivement mechanism will be created.
 
This class is used for keeping track of learners' progress in courses. When a user enrolls in a course, corresponding `Achievement` object should be created with the `user_id` and the `course_id`. Each `Achivement` object should store a list consisting of the course's checkpoints, progress array with storing `0` or `1` according to the completition status of a checkpoint. After completing certain chekpoints in a course, user should gain badges. 

### Prerequisites mentioned in [Requirements](https://github.com/bounswe/bounswe2022group3/wiki/Requirements) : 
* 2.1.2.2. Achievements
 
     * 2.1.2.2.1. System shall keep an achievement section for each user.

     * 2.1.2.2.2. System shall keep a predefined list of badges that can be earned by user.
     * 2.1.2.2.3. Users will gain a badge when one the following occurs: learner completes a course, learner completes one section of the course, lecturer opens a new course, lecturer receives a review point that is greater than 4. (see 2.1.2.7.2.)
     * 2.1.2.2.4. The system shall allow the course lecturer to add more badges to the badge list.
     * 2.1.2.2.5. Achievement page shall be updated when a user gains a badge.

## Fields
| Field| Type | Description|
| ---| ---| ---| 
| user_id | UID | ID of the user who enrolled into a course |
| course_id | UID | ID of the course the user has enrolled | 
| course_checkpoints | (content_id, int)[] | The list of tuples consisting of content_ids of the course checkpoints and their completition statuses as integers . This list is created using `course_id`. It can have any type of course content. For each content, `1` means completed and `0` means not completed.| 
| earned_badges | string[] | The list of badges (strings) the user has earned from this course |
| all_badges | (string, content_id) [] | The list of tuples storing the names of all gainable badges in this course and a `content_id` to match the badge with a checkpoint in the course. |

## Methods
| Method| Args | Return |  Description|
| ---| ---| ---| ---|
| getCourseCheckpoints | course_id : UID | content_id[] | When `Achievement` object is created, using its `course_id,` it will get a list of `content_ids` that are set as checkpoints in the course. |
| updateProgress | course_checkpoints : (content_id, int)[] | (content_id, int)[] | This method is used to update the progress of `course_checkpoints` list when a checkpoint is reached. |
| calculateTotalProgress | course_checkpoints : (content_id, int)[] | float | This method calculates the percentage of user's progress in this course. |
| updateBadges | course_checkpoints: (content_id, int)[], earned_badges : string[], all_badges : (string, content_id)[] | string[] | This method `updates` earned_badges list according to the `course_checkpoints` list and and the `all_badges`. |
| setAllBadges | all_badges : (string, content_id)[] | (string, content_id)[] | This method allows lecturers to create their own badges. |

**Deadline:** 04.04.2022 10.00

**Comment by hsuleerkul** [comment]
First Draft of Achievement class is ready and you can see it down below. 

![achievementclass](https://user-images.githubusercontent.com/56507424/161453780-e326222e-f64b-4f92-b57a-b8631bae61da.PNG)

**Comment by hsuleerkul** [comment]
After yesterday's group meeting [Meeting#7](https://github.com/bounswe/bounswe2022group3/wiki/Meeting-Notes-%237-04.04.2022) we decided that there is an overlap between `Achievement` class and `Enrollment` class. @muhammetssen and I discussed about that (#97) and we decided to change this class to `Badge` class. In order to keep things clear, I will close this issue and the progress of the `Badge` class will be kept under #99.
